runtime-sdk/src/state.rs

```diff
@@ -0,0 +1,1573 @@
+    pub(crate) fn init_local_fallback() {
+        thread_local! {
+            static BASE_STATE_INIT: RefCell<bool> = RefCell::new(false);
+        }
+
+        BASE_STATE_INIT.with(|initialized| {
+            // Initialize once per thread.
+            if *initialized.borrow() {
+                return;
+            }
+            *initialized.borrow_mut() = true;
+
+            let root = mkvs::OverlayTree::new(
+                mkvs::Tree::builder()
+                    .with_root_type(mkvs::RootType::State)
+                    .build(Box::new(mkvs::sync::NoopReadSyncer)),
+            );
+            let root = MKVSStore::new(root);
+
+            // Initialize the root state.
+            let root = State {
+                parent: None,
+                store: Some(OverlayStore::new(Box::new(root) as Box<dyn Store>)),
+                events: EventTags::new(),
+                unconditional_events: EventTags::new(),
+                messages: Vec::new(),
+                block_values: BTreeMap::new(),
+                hidden_block_values: None,
+                local_values: BTreeMap::new(),
+                rng: Some(RootRng::new(Default::default())),
+                hidden_rng: None,
+                env: Default::default(),
+                always_rollback: false,
+            };
+
+            CURRENT.with(|c| {
+                let mut current = c
+                    .try_borrow_mut()
+                    .expect("must not re-enter from with block");
+                assert!(
+                    current.is_empty(),
+                    "must have no prior states attached to local thread"
+                );
+
+                current.push(root);
+            });
+        });
+    }
+
+    /// Run a closure with the currently active state.
+    ///
+    /// # Panics
+    ///
+    /// This method will panic if called outside `CurrentState::enter` or if any transaction methods
+    /// are called from the closure.
+    pub fn with<F, R>(f: F) -> R
+    where
+        F: FnOnce(&mut State) -> R,
+    {
+        CURRENT.with(|c| {
+            let mut current_ref = c.try_borrow_mut().expect("must not re-enter with");
+            let current = current_ref.last_mut().expect("must enter context");
+
+            f(current)
+        })
+    }
+
+    /// Run a closure with the store of the currently active state.
+    ///
+    /// # Panics
+    ///
+    /// This method will panic if called outside `CurrentState::enter` or if any transaction methods
+    /// are called from the closure.
+    pub fn with_store<F, R>(f: F) -> R
+    where
+        F: FnOnce(&mut dyn Store) -> R,
+    {
+        Self::with(|state| f(state.store()))
+    }
+
+    /// Run a closure with the environment of the currently active state.
+    ///
+    /// # Panics
+    ///
+    /// This method will panic if called outside `CurrentState::enter` or if any transaction methods
+    /// are called from the closure.
+    pub fn with_env<F, R>(f: F) -> R
```

Comment:
Optional: There are still ~40 lines where this could be used instead of `with`, just search for `state.env()`.

Reply:
Should be fixed now.

Reply:
I still see ~20 lines that could be changed.